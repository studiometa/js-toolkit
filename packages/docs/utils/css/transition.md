# transition

This method implements the same logic as the [`<transition>` component](https://vuejs.org/v2/api/#transition) from Vue to apply CSS based transition to a DOM element.

## Usage

```js
import { transition } from '@studiometa/js-toolkit/utils';

// Will apply and remove the following classes sequentially:
// - fade-from
// - fade-active
// - fade-to
transition(document.body, 'fade');

// Will apply and remove the following styles sequentially:
// - opacity: 0;
// - transition: opacity 1s ease-out
transition(document.body, {
  from: { opacity: 0 },
  active: { transition: 'opacity 1s ease-out' },
});

// Will apply and remove the following classes sequentially:
// - opacity-0
// - transition duration-500
// - opacity-50
transition(document.body, {
  from: 'opacity-0',
  active: 'transition duration-500',
  to: 'opacity-50',
});

// Will apply and remove the following classes sequentially:
// - opacity-0
// - transition duration-500
// And keep the latest class at the end:
// - opacity-50
transition(
  document.body,
  {
    from: 'opacity-0',
    active: 'transition duration-500',
    to: 'opacity-50',
  },
  'keep',
);
```

### Parameters

- `element` (`HTMLElement`): the target DOM element
- `name` (`String|Object`): the name of the transition or an object of classes or styles to apply
- `endMode = 'remove'` (`String`): whether or not the transition keeps its final state or not, can be `remove` or `keep`, defaults to `remove`.

### Return value

- `Promise`: a `Promise` resolving when the transition has ended
