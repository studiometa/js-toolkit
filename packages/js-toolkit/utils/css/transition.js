import { nextFrame } from '../nextFrame.js';
import * as classes from './classes.js';
import * as styles from './styles.js';
import { isArray, isString } from '../is.js';
import { hasWindow } from '../has.js';
// eslint-disable-next-line import/extensions
import { eachElements } from './utils.js';

/** WeakMap to hold the transition instances. */
const cache = new WeakMap();

/**
 * @typedef {string|string[]|Partial<CSSStyleDeclaration>} ClassesOrStyles
 * @typedef {{ from?: ClassesOrStyles, active?: ClassesOrStyles, to?: ClassesOrStyles }} TransitionStyles
 */

/**
 * Transition class.
 */
class Transition {
  /**
   * Is a transition currently running?
   * @type {boolean}
   */
  isTransitioning = false;

  /**
   * A callback to execute when the transition ends.
   * @type {EventListenerOrEventListenerObject|null}
   */
  transitionEndHandler = null;

  /**
   * Instantiate and save the instance to the cache.
   * @param {HTMLElement} element The HTML element.
   */
  constructor(element) {
    cache.set(element, this);
  }

  /**
   * Get the transition class attached to the given element.
   * @param  {HTMLElement} element The HTML element concerned by the transition.
   * @returns {Transition}          The transition instance tied to the given element.
   */
  static getInstance(element) {
    let instance = cache.get(element);

    if (!instance) {
      instance = new this(element);
    }

    return instance;
  }
}

/**
 * Update either the classes or the styles of an element with the given method.
 *
 * @param {HTMLElement}     element         The element to update.
 * @param {ClassesOrStyles} classesOrStyles The classes or styles to apply.
 * @param {'add'|'remove'}  method          The method to use, one of `add` or `remove`.
 */
export function setClassesOrStyles(element, classesOrStyles, method = 'add') {
  if (isString(classesOrStyles) || isArray(classesOrStyles)) {
    classes[method](element, classesOrStyles);
  } else {
    styles[method](element, classesOrStyles);
  }
}

/**
 * Test if the given element has a transition duration.
 *
 * @param  {HTMLElement} element The element to test.
 * @returns {boolean}             The result of the test.
 */
function testTransition(element) {
  if (!hasWindow()) {
    return false;
  }

  const { transitionDuration } = window.getComputedStyle(element);
  return !!transitionDuration && transitionDuration !== '0s';
}

/**
 * Apply the from state.
 *
 * @param  {HTMLElement}      element         The target element.
 * @param  {TransitionStyles} classesOrStyles The classes or styles definition.
 * @returns {Promise<void>}
 */
async function start(element, classesOrStyles) {
  const trs = Transition.getInstance(element);
  trs.isTransitioning = true;
  setClassesOrStyles(element, classesOrStyles.from);
  await nextFrame();
  setClassesOrStyles(element, classesOrStyles.active);
}

/**
 * Apply the active state.
 *
 * @param  {HTMLElement}      element         The target element.
 * @param  {TransitionStyles} classesOrStyles The classes or styles definition.
 * @returns {Promise}
 */
async function next(element, classesOrStyles) {
  const hasTransition = testTransition(element);

  /* eslint-disable-next-line */
  return new Promise(async (resolve) => {
    if (hasTransition) {
      const trs = Transition.getInstance(element);
      trs.transitionEndHandler = resolve;
      element.addEventListener('transitionend', trs.transitionEndHandler, false);
    }
    setClassesOrStyles(element, classesOrStyles.from, 'remove');
    setClassesOrStyles(element, classesOrStyles.to);
    await nextFrame();
    if (!hasTransition) {
      resolve();
    }
  });
}

/**
 * Apply the final state.
 *
 * @param {HTMLElement}   element         The target element.
 * @param {string | Object} classesOrStyles The classes or styles definition.
 * @param {string}        mode            Whether to remove or keep the `to`  classes/styles.
 * @returns {void}
 */
function end(element, classesOrStyles, mode = 'remove') {
  const trs = Transition.getInstance(element);
  element.removeEventListener('transitionend', trs.transitionEndHandler, false);
  if (mode === 'remove') {
    setClassesOrStyles(element, classesOrStyles.to, 'remove');
  }
  setClassesOrStyles(element, classesOrStyles.active, 'remove');
  trs.isTransitioning = false;
  trs.transitionEndHandler = null;
}

/**
 * Manage CSS transition with class.
 *
 * This is heavily inspired by the Vue `<transition>` component
 * and the `@barba/css` package, many thanks to them!
 *
 * @param  {HTMLElement}             element The target element.
 * @param  {string|TransitionStyles} name    The name of the transition or an object with the hooks classesOrStyles.
 * @param  {string}                  endMode Whether to remove or keep the `to` classes/styles
 * @returns {Promise<void>}                   A promise resolving at the end of the transition.
 */
async function singleTransition(element, name, endMode = 'remove') {
  /** @type {TransitionStyles} */
  const classesOrStyles = isString(name)
    ? {
        from: `${name}-from`,
        active: `${name}-active`,
        to: `${name}-to`,
      }
    : {
        from: '',
        active: '',
        to: '',
        ...name,
      };

  const trs = Transition.getInstance(element);
  // End any previous transition running on the element.
  if (trs.isTransitioning) {
    end(element, classesOrStyles);
  }

  await start(element, classesOrStyles);
  await next(element, classesOrStyles);
  end(element, classesOrStyles, endMode);
  return Promise.resolve();
}

/**
 * Manage CSS transition with class.
 *
 * @param  {HTMLElement|HTMLElement[]|NodeListOf<HTMLElement>} elementOrElements The target element or elements.
 * @param  {string|TransitionStyles}            name    The name of the transition or an object with the hooks classesOrStyles.
 * @param  {string}                             endMode Whether to remove or keep the `to` classes/styles
 * @returns {Promise<void>}                             A promise resolving at the end of the transition.
 */
export default async function transition(elementOrElements, name, endMode = 'remove') {
  await Promise.all(
    eachElements(elementOrElements, (element) => singleTransition(element, name, endMode)),
  );
}
