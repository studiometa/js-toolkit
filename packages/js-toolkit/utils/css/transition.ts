import { nextFrame } from '../nextFrame.js';
import * as classes from './classes.js';
import * as styles from './styles.js';
import { isArray, isString } from '../is.js';
import { hasWindow } from '../has.js';
// eslint-disable-next-line import/extensions
import { eachElements } from './utils.js';

/** WeakMap to hold the transition instances. */
const cache = new WeakMap();

export type ClassesOrStyles = string | string[] | Partial<CSSStyleDeclaration>;

export type TransitionStyles = {
  from?: ClassesOrStyles;
  active?: ClassesOrStyles;
  to?: ClassesOrStyles;
};

/**
 * Transition class.
 */
class Transition {
  /**
   * Is a transition currently running?
   */
  isTransitioning = false;

  /**
   * A callback to execute when the transition ends.
   */
  transitionEndHandler: EventListenerOrEventListenerObject | null = null;

  /**
   * Instantiate and save the instance to the cache.
   */
  constructor(element: HTMLElement) {
    cache.set(element, this);
  }

  /**
   * Get the transition class attached to the given element.
   * @param  {HTMLElement} element The HTML element concerned by the transition.
   * @returns {Transition}          The transition instance tied to the given element.
   */
  static getInstance(element: HTMLElement): Transition {
    let instance = cache.get(element);

    if (!instance) {
      instance = new this(element);
    }

    return instance;
  }
}

/**
 * Update either the classes or the styles of an element with the given method.
 */
export function setClassesOrStyles(
  element: HTMLElement,
  classesOrStyles: ClassesOrStyles,
  method: 'add' | 'remove' = 'add',
) {
  if (isString(classesOrStyles) || isArray(classesOrStyles)) {
    classes[method](element, classesOrStyles);
  } else {
    styles[method](element, classesOrStyles);
  }
}

/**
 * Test if the given element has a transition duration.
 */
function testTransition(element: HTMLElement) {
  if (!hasWindow()) {
    return false;
  }

  const { transitionDuration } = window.getComputedStyle(element);
  return !!transitionDuration && transitionDuration !== '0s';
}

/**
 * Apply the from state.
 */
async function start(element: HTMLElement, classesOrStyles: TransitionStyles) {
  const trs = Transition.getInstance(element);
  trs.isTransitioning = true;
  setClassesOrStyles(element, classesOrStyles.from);
  await nextFrame();
  setClassesOrStyles(element, classesOrStyles.active);
}

/**
 * Apply the active state.
 */
async function next(element: HTMLElement, classesOrStyles: TransitionStyles): Promise<void> {
  const hasTransition = testTransition(element);

  /* eslint-disable-next-line */
  return new Promise(async (resolve) => {
    if (hasTransition) {
      const trs = Transition.getInstance(element);
      trs.transitionEndHandler = () => resolve();
      element.addEventListener('transitionend', trs.transitionEndHandler, false);
    }
    setClassesOrStyles(element, classesOrStyles.from, 'remove');
    setClassesOrStyles(element, classesOrStyles.to);
    await nextFrame();
    if (!hasTransition) {
      resolve();
    }
  });
}

/**
 * Apply the final state.
 */
function end(
  element: HTMLElement,
  classesOrStyles: TransitionStyles,
  mode: 'keep' | 'remove' = 'remove',
) {
  const trs = Transition.getInstance(element);
  element.removeEventListener('transitionend', trs.transitionEndHandler, false);
  if (mode === 'remove') {
    setClassesOrStyles(element, classesOrStyles.to, 'remove');
  }
  setClassesOrStyles(element, classesOrStyles.active, 'remove');
  trs.isTransitioning = false;
  trs.transitionEndHandler = null;
}

/**
 * Manage CSS transition with class.
 *
 * This is inspired by the Vue `<transition>` component
 * and the `@barba/css` package, many thanks to them!
 *
 * @param  {HTMLElement}             element The target element.
 * @param  {string|TransitionStyles} name    The name of the transition or an object with the hooks classesOrStyles.
 * @param  {string}                  endMode Whether to remove or keep the `to` classes/styles
 * @returns {Promise<void>}                   A promise resolving at the end of the transition.
 */
async function singleTransition(
  element: HTMLElement,
  name: string | TransitionStyles,
  endMode: 'keep' | 'remove' = 'remove',
): Promise<void> {
  /** @type {TransitionStyles} */
  const classesOrStyles = isString(name)
    ? {
        from: `${name}-from`,
        active: `${name}-active`,
        to: `${name}-to`,
      }
    : {
        from: '',
        active: '',
        to: '',
        ...name,
      };

  const trs = Transition.getInstance(element);
  // End any previous transition running on the element.
  if (trs.isTransitioning) {
    end(element, classesOrStyles);
  }

  await start(element, classesOrStyles);
  await next(element, classesOrStyles);
  end(element, classesOrStyles, endMode);
  return Promise.resolve();
}

/**
 * Manage CSS transition with class.
 *
 * @param  {HTMLElement|HTMLElement[]|NodeListOf<HTMLElement>} elementOrElements The target element or elements.
 * @param  {string|TransitionStyles}                           name              The name of the transition or an object with the hooks classesOrStyles.
 * @param  {string}                                            endMode           Whether to remove or keep the `to` classes/styles
 * @returns {Promise<void>}                                                      A promise resolving at the end of the transition.
 */
export default async function transition(
  elementOrElements: HTMLElement | HTMLElement[] | NodeListOf<HTMLElement>,
  name: string | TransitionStyles,
  endMode: 'keep' | 'remove' = 'remove',
) {
  await Promise.all(
    eachElements(elementOrElements, (element) => singleTransition(element, name, endMode)),
  );
}
