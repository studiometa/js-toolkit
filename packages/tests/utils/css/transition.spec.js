import { jest } from '@jest/globals';
import transition from '@studiometa/js-toolkit/utils/css/transition';

describe('transition method', () => {
  let el;
  let spyAdd;
  let spyRemove;
  let spyStyle;

  beforeEach(() => {
    jest.useRealTimers();
    document.body.innerHTML = `
      <div class="w-16 h-16 m-10 bg-black"></div>
    `;
    el = document.body.firstElementChild;
    spyAdd = jest.spyOn(el.classList, 'add');
    spyRemove = jest.spyOn(el.classList, 'remove');
    spyStyle = jest.spyOn(el.style, 'opacity', 'set');
  });

  it('should work server side', async () => {
    // Mock window === undefined
    // @see https://stackoverflow.com/a/56999581
    const windowSpy = jest.spyOn(globalThis, 'window', 'get');
    windowSpy.mockImplementation(() => undefined);
    await transition(el, 'name');
    expect(spyAdd).toHaveBeenCalledWith('name-from');
    windowSpy.mockRestore();
  });

  it('should work with a string parameter', async () => {
    el.style.transitionDuration = '0.1s';
    setTimeout(() => {
      el.dispatchEvent(new CustomEvent('transitionend'));
    }, 100);
    await transition(el, 'name');
    expect(spyAdd).toHaveBeenNthCalledWith(1, 'name-from');
    expect(spyAdd).toHaveBeenNthCalledWith(2, 'name-active');
    expect(spyAdd).toHaveBeenNthCalledWith(3, 'name-to');
    expect(spyRemove).toHaveBeenNthCalledWith(1, 'name-from');
    expect(spyRemove).toHaveBeenNthCalledWith(2, 'name-to');
    expect(spyRemove).toHaveBeenNthCalledWith(3, 'name-active');
  });

  it('should work with an object parameter', async () => {
    await transition(el, {
      from: { opacity: '0' },
      active: 'transition duration-500',
      to: 'transform scale-50',
    });
    expect(spyStyle.mock.calls).toMatchSnapshot();
    expect(spyAdd.mock.calls).toMatchSnapshot();
    expect(spyRemove.mock.calls).toMatchSnapshot();
  });

  it('should stop any previous transition', async () => {
    el.style.transitionDuration = '1s';
    setTimeout(() => {
      el.dispatchEvent(new CustomEvent('transitionend'));
    }, 100);
    transition(el, 'name');
    await transition(el, 'name');
    expect(spyRemove).toHaveBeenNthCalledWith(1, 'name-to');
    expect(spyRemove).toHaveBeenNthCalledWith(2, 'name-active');
  });
});
