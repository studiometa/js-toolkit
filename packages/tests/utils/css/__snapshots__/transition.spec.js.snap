// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transition method should work with an object parameter 1`] = `
[
  [
    "0",
  ],
  [
    "",
  ],
]
`;

exports[`transition method should work with an object parameter 2`] = `
[
  [
    "transition",
    "duration-500",
  ],
  [
    "transform",
    "scale-50",
  ],
]
`;

exports[`transition method should work with an object parameter 3`] = `
[
  [
    "transform",
    "scale-50",
  ],
  [
    "transition",
    "duration-500",
  ],
]
`;
